### PR TITLE
Don't mangle production builds

### PR DESCRIPTION
**Pull request by swankjesse**
We want to remove dead code, but not much more. There's not much benefit to minifying code before we send it into the QuickJS compiler.

### CODE REVIEW

**Comment by swankjesse** [comment]
Turns out there’s a cost to this!

For a non-trivial application, this causes a 42% increase in the download size.

```
                       NO DEBUG DATA    DEBUG DATA    DELTA
        bytecode size:     2,856,719     3,871,128     +35%
gzipped bytecode size:       941,474     1,346,230     +42%
```

I’ll see if I can tighten that up...

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      |      },
+      |      mangle: false,
+      |      format: {
+      |        beautify: true,
+      |        braces: true,
+      |      }
+      |    },
+      |  }),
+      |];
+      """.trimMargin(),
```
Comment (JakeWharton) [inline zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt:65]:
```suggestion
      |""".trimMargin(),
```
Trailing newline always a good idea

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    )
+  }
+
+  companion object {
+    val generatedFilePath = "webpack.config.d/generated-zipline-webpack-config.js"
```
Comment (JakeWharton) [inline zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt:70]: I thought these had to be prefixed with a number like `23-whevs.js`. Maybe that's just a convention for ordering.

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt [hunk 3]
```diff
@@ -67,17 +67,21 @@ class ZiplinePlugin : KotlinCompilerPluginSupportPlugin {
     }
 
     target.tasks.withType(KotlinWebpack::class.java) { kotlinWebpack ->
```
Comment (JakeWharton) [inline zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt:69]: Unrelated to this change, but I'm pretty sure this causes all tasks to be configured. Gradle docs will have the lazy pattern. Separate PR for that...

File: zipline-testing/build.gradle.kts [hunk 4]
```diff
@@ -68,7 +68,6 @@ kotlin {
 }
 
 tasks {
-  // TODO upstream this to ZiplinePlugin
   withType<KotlinJsCompile>().all {
     kotlinOptions {
       freeCompilerArgs += listOf("-Xir-per-module")
```
Comment (JakeWharton) [inline zipline-testing/build.gradle.kts:73]: Delete it all. This is the default!

**Review by JakeWharton — APPROVED** [review]

**Comment by swankjesse** [comment]
More analysis of file size costs:

```
                                    .zipline size        gzipped size
A NO DEBUG DATA                    2,856,719             941,474
B MANGLE CLASS + FUNCTION NAMES    3,400,044 (+19%)    1,193,241 (+27%)
C OMIT INSTRUCTION LINE NUMBERS    3,690,174 (+29%)    1,218,497 (+29%)
D TRUNCATE FILE PATH               3,642,791 (+28%)    1,338,461 (+42%)
E (B + C)                          3,218,986 (+13%)    1,064,318 (+13%)
F FULL DEBUG DATA                  3,871,128 (+35%)    1,346,230 (+42%)
```

Item B changes the Terser config to enable mangling, which will prevent true function names from occurring in stacktraces:
```
 new TerserPlugin({
   terserOptions: {
     compress: {
       sequences: false,
     },
     mangle: {
     },
     format: {
       beautify: true,
       braces: true,
     }
   },
 }),
```

Item C clears the pc2line entry from the bytecode, which will prevent true line numbers from occurring in stacktraces.

I think our best option for production use is B+C which still gives us bad stacktraces (file names only, no line numbers or function names), but with dramatically lower cost.

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt (hunk 1 above)
Comment (swankjesse) [reply zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt:65]: ktlint really has no taste.

```
> The following files had format violations:
      zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt
          @@ -60,7 +60,8 @@
           ······|····},
           ······|··}),
           ······|];
          -······|""".trimMargin(),
          +······|
          +""".trimMargin(),
           ····)
           ··}
```

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt (hunk 2 above)
Comment (swankjesse) [reply zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/WriteWebpackConfigTask.kt:70]: Yep. I don’t expect that to be useful for this use.

File: zipline-testing/build.gradle.kts (hunk 4 above)
Comment (swankjesse) [reply zipline-testing/build.gradle.kts:73]: Follow-up coming up.

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt (hunk 3 above)
Comment (swankjesse) [reply zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt:69]: Follow-up coming up.

**Comment by JakeWharton** [comment]
The source map doesn't/can't help at all here?

**Comment by JakeWharton** [comment]
Or what if we rewrote/appended to the R8 mapping file so the short, mangled names would get deobfuscated by crash reporting tools entirely on the server-side?

**Comment by JakeWharton** [comment]
Can be future investigations

**Comment by swankjesse** [comment]
Yep, we could do it server-side, though we’ll introduce some complexity to support that. The biggest challenge I see is that our QuickJS stack frames don’t consistently have a class+function, and I assume that’s how deobfuscators are keyed.

**Comment by JakeWharton** [comment]
They are, but since R8 might put everything into a single class we only need to use a fake one to get that working.

File: zipline-testing/build.gradle.kts (hunk 4 above)
Comment (swankjesse) [reply zipline-testing/build.gradle.kts:73]: https://github.com/cashapp/zipline/pull/1179

File: zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt (hunk 3 above)
Comment (swankjesse) [reply zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplinePlugin.kt:69]: https://github.com/cashapp/zipline/issues/1180

**Pull request merged**